samcli/hook_packages/terraform/hooks/prepare/resource_linking.py

```diff
@@ -118,14 +191,66 @@ def _handle_linking(self, source_tf_resource: TFResource, cfn_source_resources:
                 "There are destination resources defined for for the source resource %s",
                 source_tf_resource.full_address,
             )
-        else:
-            self._update_mapped_parent_resource_with_resolved_child_resources(cfn_source_resources, dest_resources)
+        for cfn_resource in cfn_resources:
+            self._resource_pair.cfn_resource_update_call_back_function(cfn_resource, dest_resources)
+
+    def _link_using_linking_fields(self, cfn_resource: Dict) -> None:
+        """
+        Depends on that all the child resources of the source resource are applied, and so we do not need to traverse
+        the terraform configuration to define the destination resource. We will depend on the actual values of the
+        linking fields, and find the destination resource that has the same value.
+
+        Parameters
+        ----------
+        cfn_source_resource: Dict
+            A mapped CFN source resource
+        """
+        # get the constant values of the linking field from the cfn_resource
+        values = cfn_resource.get("Properties", {}).get(self._resource_pair.cfn_link_field_name)
+
+        LOG.debug(
+            "Link the source resource %s using linking property %s that has the value %s",
+            cfn_resource,
+            self._resource_pair.cfn_link_field_name,
+            values,
+        )
+
+        # some resources can be linked to only one child resource like rest apis.
+        # make the resource values as a list to make processing easier.
+        if not isinstance(values, List):
+            values = [values]
+
+        # build map between the destination linking field property values, and resources' logical ids
+        child_resources_linking_attributes_logical_id_mapping = {
+            self._resource_pair.destination_resource_tf.get(logical_id, {})
+            .get("values", {})
+            .get(self._resource_pair.tf_destination_attribute_name): logical_id
+            for logical_id in self._resource_pair.destination_resource_tf.keys()
+            if self._resource_pair.destination_resource_tf.get(logical_id, {})
+            .get("values", {})
+            .get(self._resource_pair.tf_destination_attribute_name)
+        }
```

Comment:
Not completely following this. If we know that the reference is a constant value arn, won't 
```
self._resource_pair.destination_resource_tf.get(logical_id, {})
            .get("values", {})
            .get(self._resource_pair.tf_destination_attribute_name)
```
also yeild an arn? But we later create a `LogicalIdReference` if there is something in this map.

Reply:
let me rewrite it using loops to be clear. I am creating a map between Layers `ARN` values (for example) and the Layers logical Ids, then use this map to search for the layers that has the same `ARN` values exist in the lambda functions.layers

Reply:
Done

Reply:
I think the part that's confusing me a little is this 
```
dest_resources = [
      LogicalIdReference(child_resources_linking_attributes_logical_id_mapping[value])
      if value in child_resources_linking_attributes_logical_id_mapping
      else ExistingResourceReference(value)
      for value in values
]
```
I guess we can have a list that is a combination of `LogicalIdReferences` and `ExistingReferences`, but I thought that if 

```
linking_attribute_value = destination_resource.get("values", {}).get(
      self._resource_pair.tf_destination_attribute_name
)
```
is not none it has to be an arn. So when we create `dest_resources` it will only create `LogicalIdReference`.

Reply:
Discussed offline. This value should in fact be a logical id reference even if we have a static value for better supporting `sam local` commands. If the value is a static arn, AWS SAM CLI will try to pull it instead of using the locally defined resource.